convert-hf-to-gguf.py

```diff
@@ -91,10 +92,27 @@ def __init__(self, dir_model: Path, ftype: gguf.LlamaFileType, fname_out: Path,
             else:
                 logger.info(f"choosing --outtype bf16 from first tensor type ({first_tensor.dtype})")
                 self.ftype = gguf.LlamaFileType.MOSTLY_BF16
-        ftype_up: str = self.ftype.name.partition("_")[2].upper()
-        ftype_lw: str = ftype_up.lower()
+
+        # Generate default filename based on model specification and available metadata
+        version_string = None # TODO: Add metadata support
+        expert_count = self.hparams["num_local_experts"] if "num_local_experts" in self.hparams else None
+        encodingScheme = {
+            gguf.LlamaFileType.ALL_F32     : "F32",
+            gguf.LlamaFileType.MOSTLY_F16  : "F16",
+            gguf.LlamaFileType.MOSTLY_BF16 : "BF16",
+            gguf.LlamaFileType.MOSTLY_Q8_0 : "Q8_0",
+        }[self.ftype]
+        self.fname_default = f"{gguf.naming_convention(dir_model.name, version_string, expert_count, self.parameter_count(), encodingScheme)}"
+
+        # Filename Output
+        if fname_out is not None:
+            # custom defined filename and path was provided
+            self.fname_out = fname_out
+        else:
+            # output in the same directory as the model by default
+            self.fname_out = dir_model.parent / self.fname_default
+
         # allow templating the file name with the output ftype, useful with the "auto" ftype
-        self.fname_out = fname_out.parent / fname_out.name.format(ftype_lw, outtype=ftype_lw, ftype=ftype_lw, OUTTYPE=ftype_up, FTYPE=ftype_up)
```

Comment:
I think there should still be a way to get custom name templates with either uppercased or lowercased type.

From the non-removal of the comment above, I guess you did not intend to remove this feature.

Reply:
I mean, when a name is provided, type templates should be possible as before. Otherwise, custom names can't contain the real quantization type with `--outtype auto`.

This was there to allow this, and was a feature requested in <https://github.com/ggerganov/llama.cpp/pull/7158#discussion_r1596930517>.